### DISCUSSION TITLE
How we can eliminate(remove) middleware in test mode (test api)?

### DISCUSSION THREAD

**Discussion by pouryajafarzadeh** [fastapi/fastapi#7098] [Questions] [answered]
In fact, override the main middleware with another middleware (bypass the middleware) in a test mode.

**Comment by Mause** [comment]
Have you tried not adding the middleware in the first place? Then you don't have to remove it

**Comment by pouryajafarzadeh** [comment]
Yes. I could do that by removing middleware in the main app code, but I want to eliminate middleware in the test mode, when we are using:

`from fastapi.testclient import TestClient`
`client =  TestClient(app)`

**Comment by falkben** [comment, marked as answer]
yes, i think it's probably easier to never attach the middleware rather than try and remove it.

you could create the app in the test, and just leave off the middleware, just grabbing the router(s) (assuming you are using routers). that way your app is in main or wherever and the router is... wherever it is, but doesn't get any middleware attached to it.

or you could create a function to create the app which selectively mounts middleware  based on environment, so that you can still import and test the app w/o the middleware. something like the function proposed in [this comment](https://github.com/tiangolo/fastapi/issues/582#issuecomment-537015682).

curious though... wouldn't you want your tests to use the same middleware as the app does in production?

**Comment by pouryajafarzadeh** [comment]
Thank you. I find out your proposed solution.

**Comment by tiangolo** [comment]
Thanks for the help here @Mause  and @falkben  ! :clap: :bow: 

Thanks for reporting back and closing the issue @pouryajafarzadeh  :+1:

**Comment by mdryden** [comment]
I hope you don't mind a post on this old thread, but I found this issue while I was searching for solution to the same problem.  I didn't like the linked solution, but ended up working out my own.  Perhaps this will be useful to someone else:

```
test_client = TestClient(app)

app.user_middleware.clear()
app.middleware_stack = app.build_middleware_stack()
```

This seems to do what I (and probably the OP) was after.  In my case, I wanted to disable the middleware to avoid dealing with an API key on requests.  Since user_middleware is a List, I suppose you could also selectively remove one or more middleware components without removing everything as well.

**Comment by dmanthing** [comment]
```
def remove_middleware(app: FastAPI, target: str) -> FastAPI:
    new_middlewares: list[Middleware] = []
    for middleware in app.user_middleware:
        if not middleware.cls.__name__ == target:
            new_middlewares.append(middleware)
    app.user_middleware = new_middlewares
    app.middleware_stack = app.build_middleware_stack()
    return app
```
    
just pass the middleware class name as the `target` param :)

**Comment by rajshahdev** [comment]
> ```
>     new_middlewares: list[Middleware] = []
```
from where should I import Middleware @dmanthing

**Comment by dmanthing** [comment]
> > ```
> >     new_middlewares: list[Middleware] = []
> > ```
> 
> ```
> from where should I import Middleware @dmanthing 
> ```

```
from fastapi import FastAPI
from starlette.middleware import Middleware


def remove_middleware(app: FastAPI, target: str) -> FastAPI:
    new_middlewares: list[Middleware] = []
    for middleware in app.user_middleware:
        if not middleware.cls.__name__ == target:
            new_middlewares.append(middleware)
    app.user_middleware = new_middlewares
    app.middleware_stack = app.build_middleware_stack()
    return app
```

**Reply by allen-munsch** [reply]
Thank you! very nice! I made an addition for middlewares which are added via decorator:

For example on:

```
@webapp.middleware("http")
async def login_required_middleware(request: Request, call_next):
    return await login_required(request, call_next)
```

I added a check for a dispatch, it seemed to work:

```
def remove_middleware(app: FastAPI, target: str) -> FastAPI:
    new_middlewares: list[Middleware] = []
    for middleware in app.user_middleware:
        # this might need to use kwargs instead?
        # dispatcher = getattr(middleware, 'kwargs', {}).get('dispatch', object)
        dispatcher = getattr(middleware, 'options', {}).get('dispatch', object)
        if target not in [middleware.cls.__name__, getattr(dispatcher, '__name__', 'NONE')]:
            new_middlewares.append(middleware)
    app.user_middleware = new_middlewares
    app.middleware_stack = app.build_middleware_stack()
    return app
```

Usage example:

```
def get_webapp():
    from myapp.serve import webapp
    webapp = remove_middleware(webapp, 'login_required_middleware')
    webapp = remove_middleware(webapp, 'AuthenticationMiddleware')
    return webapp
    
 def my_test():
     from starlette.testclient import TestClient
     client = TestClient(get_webapp())
     client.post('/login-required-endpoint', json={'hello': 'world'})
```
